## Instagram_images

In [583]:
import pandas as pd
from os import listdir
import json
import re
from pprint import pprint
from tqdm import tqdm_notebook as tqdm
import operator
tqdm().pandas()
from shutil import copyfile

In [455]:
def get_imageid_link_lists(filename):
    df = pd.read_csv(filename, sep=' ',names=['image_id','links'])
    return df['image_id'].tolist(),df['links'].tolist()

In [456]:
def get_filenames_in_folder(foldername):
    return [i.replace('.jpg','') for i in listdir(foldername)]

In [457]:
def show_image_id_txt_file_map(filename,foldername):
    image_id,links = get_imageid_link_lists(filename)
    image_ids_folder = get_filenames_in_folder(foldername)
    print(f"FOLDER = {foldername.split('/')[-2]} and FILE= {filename.split('/')[-1]}")
    print(f"Total id and links present in txt file are {len(image_id),len(links)}")
    print(f"Total unique id and links present in txt file are {len(set(image_id)),len(set(links))}")
    print(f"Total images present in folder {len(image_ids_folder)}")
    print(f"Total unique images present in folder {len(set(image_ids_folder))}")
    print(f"Number of images present in both sets {len(set(image_ids_folder)&set(image_id))}")
    print(f"Number of IDs present in file but not in folder {len(set(image_id)-set(image_ids_folder))}")
    print(f"Number of IDs present in folder but not in file {len(set(image_ids_folder)-set(image_id))}")
    return image_id,image_ids_folder

In [458]:
image_id_0_10_file,image_id_0_10_folder = show_image_id_txt_file_map('data/instagram_images/namesall_info_media_ids_to_collect_0plus_to_10.txt','data/instagram_images/images_0_10/')

FOLDER = images_0_10 and FILE= namesall_info_media_ids_to_collect_0plus_to_10.txt
Total id and links present in txt file are (270, 270)
Total unique id and links present in txt file are (270, 270)
Total images present in folder 276
Total unique images present in folder 276
Number of images present in both sets 267
Number of IDs present in file but not in folder 3
Number of IDs present in folder but not in file 9


In [459]:
image_id_10_40_file,image_id_10_40_folder = show_image_id_txt_file_map('data/instagram_images/namesall_info_media_ids_to_collect_10plus_to_40.txt','data/instagram_images/images_10_40/')

FOLDER = images_10_40 and FILE= namesall_info_media_ids_to_collect_10plus_to_40.txt
Total id and links present in txt file are (835, 835)
Total unique id and links present in txt file are (835, 835)
Total images present in folder 987
Total unique images present in folder 987
Number of images present in both sets 831
Number of IDs present in file but not in folder 4
Number of IDs present in folder but not in file 156


In [460]:
image_id_40_plus_file,image_id_40_plus_folder = show_image_id_txt_file_map('data/instagram_images/namesall_info_media_ids_to_collect_40plus.txt','data/instagram_images/images_40plus//')

FOLDER =  and FILE= namesall_info_media_ids_to_collect_40plus.txt
Total id and links present in txt file are (816, 816)
Total unique id and links present in txt file are (816, 816)
Total images present in folder 1000
Total unique images present in folder 1000
Number of images present in both sets 809
Number of IDs present in file but not in folder 7
Number of IDs present in folder but not in file 191


In [461]:
all_ids_file = image_id_0_10_file+image_id_10_40_file+image_id_40_plus_file
all_ids_folder = image_id_0_10_folder+image_id_10_40_folder+image_id_40_plus_folder

In [462]:
print(f"Total image ids file {len(all_ids_file)}")
print(f"Total unique image ids file {len(set(all_ids_file))}")
print(f"Total images folder {len(all_ids_folder)}")
print(f"Total unique images folder {len((all_ids_folder))}")

print(f"Number of images present in both sets {len(set(all_ids_file)&set(all_ids_folder))}")
print(f"Number of IDs present in file but not in folder {len(set(all_ids_file)-set(all_ids_folder))}")
print(f"Number of IDs present in folder but not in file {len(set(all_ids_folder)-set(all_ids_file))}")

Total image ids file 1921
Total unique image ids file 1921
Total images folder 2263
Total unique images folder 2263
Number of images present in both sets 1907
Number of IDs present in file but not in folder 14
Number of IDs present in folder but not in file 356


---
**NOTE**

Looks like that these files are a map between image name (id) and source link.

---

## Image enities labels

In [463]:
def get_links_and_entites(filename):
    df = pd.read_csv(filename, sep=',',names=["image_id","label_image"])
    return df['image_id'].tolist(),df['label_image'].tolist()

In [70]:
links,_ = get_links_and_entites("data/Insta_labeled_data/image_labels_0plus_to_10_.csv")
len(links)

278

In [71]:
links,_ = get_links_and_entites("data/Insta_labeled_data/image_labels_10plus_to_40.csv")
len(links)

1032

In [72]:
links,_ = get_links_and_entites("data/Insta_labeled_data/image_labels_40plus.csv")
len(links)

953

---
**NOTE**

* Entites label of each image.
* Mapping is done using the links.
* Answers of multiple annotators are given.
---

## Labeled images

In [73]:
df = pd.read_csv('data/Insta_labeled_data/labeled_0plus_to_10__full.csv', encoding='latin-1')
len(df['img_url']),len(set(df['img_url']))

(1390, 278)

In [74]:
df = pd.read_csv('data/Insta_labeled_data/labeled_10plus_to_40_full.csv', encoding='latin-1')
len(df['img_url']),len(set(df['img_url']))

(5196, 1031)

In [75]:
df = pd.read_csv('data/Insta_labeled_data/labeled_40plus_full.csv', encoding='latin-1')
len(df['img_url']),len(set(df['img_url']))

/home/hitkul/anaconda3/envs/nii_project/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (4,216,217) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(6104, 935)

In [76]:
df.head()

_unit_id         _created_at  _golden         _id _missed  \
0  649719574  1/17/2015 19:37:16    False  1531470430     NaN   
1  649719574  1/17/2015 20:40:57    False  1531507988     NaN   
2  649719574  1/17/2015 21:54:57    False  1531549350     NaN   
3  649719574  1/18/2015 00:11:28    False  1531628330     NaN   
4  649719574  1/18/2015 02:21:20    False  1531708189     NaN   

          _started_at  _tainted   _channel  _trust  _worker_id      ...        \
0  1/17/2015 19:33:34     False  clixsense  1.0000    26831570      ...         
1  1/17/2015 20:33:40     False      elite  0.9800    30609759      ...         
2  1/17/2015 21:53:10     False    prodege  0.9200     3696575      ...         
3  1/18/2015 00:07:23     False     neodev  0.9200    29750518      ...         
4  1/18/2015 02:16:41     False  clixsense  0.8889     7398771      ...         

   clmn97  clmn98  clmn99                                          cptn_time  \
0  empety  empety  empety  <font color="#0066CC">offthis_</font>   erase ...   
1  empety  empety  empety  <font color="#0066CC">offthis_</font>   erase ...   
2  empety  empety  empety  <font color="#0066CC">offthis_</font>   erase ...   
3  empety  empety  empety  <font color="#0066CC">offthis_</font>   erase ...   
4  empety  empety  empety  <font color="#0066CC">offthis_</font>   erase ...   

                                             img_url      likes  \
0  http://scontent-b.cdninstagram.com/hphotos-xap...  196 likes   
1  http://scontent-b.cdninstagram.com/hphotos-xap...  196 likes   
2  http://scontent-b.cdninstagram.com/hphotos-xap...  196 likes   
3  http://scontent-b.cdninstagram.com/hphotos-xap...  196 likes   
4  http://scontent-b.cdninstagram.com/hphotos-xap...  196 likes   

                           owner_cmnt                                owner_id  \
0  Media posted at 012-02-11 23:04:28  <font color="#0066CC">nailsfv</font>     
1  Media posted at 012-02-11 23:04:28  <font color="#0066CC">nailsfv</font>     
2  Media posted at 012-02-11 23:04:28  <font color="#0066CC">nailsfv</font>     
3  Media posted at 012-02-11 23:04:28  <font color="#0066CC">nailsfv</font>     
4  Media posted at 012-02-11 23:04:28  <font color="#0066CC">nailsfv</font>     

  question1_gold question2_gold  
0            NaN            NaN  
1            NaN            NaN  
2            NaN            NaN  
3            NaN            NaN  
4            NaN            NaN  

[5 rows x 218 columns]

---
**NOTE**
+ *_unit_id* - unique id for each image
+ Each row is answer of one annotator
+ Other information about workers present 
+ *question1* is Aggression or not Aggrestion
+ *question2* is Bullying or Not bullying
+ Comments on the image are also given
+ Image links used as a key
+ Looks like owner comment is the caption
---

## Session Label

In [77]:
df_0_10 = pd.read_csv('data/Insta_labeled_data/sessions_0plus_to_10_metadata.csv', encoding='latin-1')
len(df_0_10['img_url']),len(set(df_0_10['img_url']))

(278, 278)

In [78]:
df_10_40 = pd.read_csv('data/Insta_labeled_data/sessions_10plus_to_40_metadata.csv', encoding='latin-1')
len(df_10_40['img_url']),len(set(df_10_40['img_url']))

(1018, 1017)

In [79]:
df_40 = pd.read_csv('data/Insta_labeled_data/sessions_40plus_metadata.csv', encoding='latin-1')
len(df_40['img_url']),len(set(df_40['img_url']))

(922, 921)

In [80]:
df_40.head()

_unit_id  _golden _unit_state  _trusted_judgments _last_judgment_at  \
0  649719574    False   finalized                   5      1/18/15 2:21   
1  649719575    False   finalized                   5      1/17/15 7:45   
2  649719576    False   finalized                   5     1/17/15 17:15   
3  649719577    False   finalized                   5      1/18/15 3:04   
4  649719578    False   finalized                   5     1/17/15 12:10   

    question1  question1:confidence question2  question2:confidence  \
0  aggression                1.0000  bullying                1.0000   
1  aggression                1.0000  bullying                0.6227   
2  aggression                1.0000  bullying                1.0000   
3  aggression                0.7876  bullying                0.5923   
4  aggression                1.0000  bullying                1.0000   

                                               clmn1      ...       \
0  <font color="#0066CC">margretricee</font>   Ba...      ...        
1  <font color="#0066CC">mightyzulu</font>   Damn...      ...        
2  <font color="#0066CC">wayne_dizzle818</font>  ...      ...        
3  <font color="#0066CC">samantha_nora</font>   S...      ...        
4  <font color="#0066CC">itsmyrayeraye</font>   <...      ...        

                                             img_url       likes  \
0  http://scontent-b.cdninstagram.com/hphotos-xap...   196 likes   
1  http://scontent-a.cdninstagram.com/hphotos-xpa...   570 likes   
2  http://scontent-a.cdninstagram.com/hphotos-xfa...    44 likes   
3  http://scontent-b.cdninstagram.com/hphotos-xpf...  1381 likes   
4  http://scontent-b.cdninstagram.com/hphotos-xpf...  1936 likes   

                                          owner_cmnt  \
0                 Media posted at 012-02-11 23:04:28   
1                                new bong mac got me   
2   So I find out that I am missing some of my Fu...   
3                                       Beautiful <3   
4                                            Tragedy   

                                         owner_id shared media followed_by  \
0          <font color="#0066CC">nailsfv</font>            189       95224   
1          <font color="#0066CC">treejtv</font>           1341       61385   
2  <font color="#0066CC">danielrochatat2</font>            945       50710   
3    <font color="#0066CC">chelseahouska</font>            711     1905124   
4    <font color="#0066CC">andreaschoice</font>            443      532939   

  follows         id cyberaggression cyberbullying  
0    6117  649719574               5             5  
1     193  649719575               5             3  
2    1071  649719576               5             5  
3     272  649719577               4             3  
4     427  649719578               5             5  

[5 rows x 215 columns]

---
**NOTE**
+ *_unit_id* is unique id for every image
+ *_trusted_judgments* = Number of trusted annotators
+ *question1* = Aggression or not
+ *question2* = Bullying or not
+ Metadata like shares, like and comments is also present 
+ Cyberaggression and cyberbullying is the count of how many annotators called it that
+ *image_url* can be used for mapping
---

# Final Dataset

Concatinating all three frames

In [82]:
final_df = pd.concat([df_0_10,df_10_40,df_40], ignore_index=True,sort=True) 

In [88]:
final_df.head()

_golden _last_judgment_at  _trusted_judgments   _unit_id _unit_state  \
0    False      4/11/15 0:40                   5  702714440   finalized   
1    False      4/11/15 0:37                   5  702714441   finalized   
2    False      4/11/15 2:17                   5  702714442   finalized   
3    False      4/11/15 2:09                   5  702714443   finalized   
4    False     4/10/15 23:41                   5  702714444   finalized   

                                               clmn1  \
0  <font color="#0066CC">lynnseyerinalexandra</fo...   
1  <font color="#0066CC">britterkritter</font>   ...   
2  <font color="#0066CC">mariamagv</font>   where...   
3  <font color="#0066CC">gladlychadly</font>   _Ž...   
4  <font color="#0066CC">fivefour</font>   <font ...   

                                              clmn10  \
0  <font color="#0066CC">niazki.lam</font>   larr...   
1  <font color="#0066CC">kyleepickering</font>   ...   
2  <font color="#0066CC">julianarosa67</font>   F...   
3  <font color="#0066CC">crash_edisy</font>   <fo...   
4  <font color="#0066CC">brohr285</font>   Still ...   

                                             clmn100  \
0                                             empety   
1                                             empety   
2  <font color="#0066CC">oopshiharry</font>   Can...   
3                                             empety   
4  <font color="#0066CC">buddhaplaysbass</font>  ...   

                                             clmn101  \
0                                             empety   
1                                             empety   
2  <font color="#0066CC">nuriaat</font>   <font c...   
3                                             empety   
4  <font color="#0066CC">tjbyerly</font>   <font ...   

                                             clmn102    ...     \
0                                             empety    ...      
1                                             empety    ...      
2  <font color="#0066CC">agusjalo</font>   This i...    ...      
3                                             empety    ...      
4  <font color="#0066CC">fivefour</font>   <font ...    ...      

                                             img_url             likes  \
0  https://scontent.cdninstagram.com/hphotos-xaf1...    4313\r\r likes   
1  https://scontent.cdninstagram.com/hphotos-xap1...    2102\r\r likes   
2  https://scontent.cdninstagram.com/hphotos-xap1...  108939\r\r likes   
3  https://scontent.cdninstagram.com/hphotos-xaf1...    6441\r\r likes   
4  https://scontent.cdninstagram.com/hphotos-xap1...    1153\r\r likes   

                                          owner_cmnt  \
0  gosh we love these five idiots so much __ for...   
1  Anyone see the Pepsi halftime show commercial ...   
2  Today I ate cr_pes and bought a drawing of ch...   
3  New works in progress for an upcoming show at ...   
4  Due to overwhelming response, we have went ahe...   

                                      owner_id   question1  \
0     <font color="#0066CC">maliksbby</font>       noneAgg   
1  <font color="#0066CC">witneycarson</font>       noneAgg   
2   <font color="#0066CC">namelessgem</font>       noneAgg   
3          <font color="#0066CC">kaws</font>    aggression   
4      <font color="#0066CC">fivefour</font>       noneAgg   

  question1:confidence question2 question2:confidence shared media     label  
0               0.8124   noneBll               1.0000         1477    normal  
1               1.0000   noneBll               1.0000          761    normal  
2               0.8056   noneBll               0.8056          698    normal  
3               0.5982  bullying               0.5982         1257  bullying  
4               1.0000   noneBll               1.0000         2052    normal  

[5 rows x 216 columns]

Getting final label for each row

In [89]:
def get_final_label(row):
    if row['question1']=='noneAgg' and row['question2']=='noneBll':
        return 'normal'
    elif row['question1']=='aggression' and row['question2']=='noneBll':
        return 'aggression'
    else:
        return 'bullying'

In [90]:
final_df['label'] = final_df.apply (lambda row: get_final_label(row),axis=1)

In [91]:
final_df.head()

_golden _last_judgment_at  _trusted_judgments   _unit_id _unit_state  \
0    False      4/11/15 0:40                   5  702714440   finalized   
1    False      4/11/15 0:37                   5  702714441   finalized   
2    False      4/11/15 2:17                   5  702714442   finalized   
3    False      4/11/15 2:09                   5  702714443   finalized   
4    False     4/10/15 23:41                   5  702714444   finalized   

                                               clmn1  \
0  <font color="#0066CC">lynnseyerinalexandra</fo...   
1  <font color="#0066CC">britterkritter</font>   ...   
2  <font color="#0066CC">mariamagv</font>   where...   
3  <font color="#0066CC">gladlychadly</font>   _Ž...   
4  <font color="#0066CC">fivefour</font>   <font ...   

                                              clmn10  \
0  <font color="#0066CC">niazki.lam</font>   larr...   
1  <font color="#0066CC">kyleepickering</font>   ...   
2  <font color="#0066CC">julianarosa67</font>   F...   
3  <font color="#0066CC">crash_edisy</font>   <fo...   
4  <font color="#0066CC">brohr285</font>   Still ...   

                                             clmn100  \
0                                             empety   
1                                             empety   
2  <font color="#0066CC">oopshiharry</font>   Can...   
3                                             empety   
4  <font color="#0066CC">buddhaplaysbass</font>  ...   

                                             clmn101  \
0                                             empety   
1                                             empety   
2  <font color="#0066CC">nuriaat</font>   <font c...   
3                                             empety   
4  <font color="#0066CC">tjbyerly</font>   <font ...   

                                             clmn102    ...     \
0                                             empety    ...      
1                                             empety    ...      
2  <font color="#0066CC">agusjalo</font>   This i...    ...      
3                                             empety    ...      
4  <font color="#0066CC">fivefour</font>   <font ...    ...      

                                             img_url             likes  \
0  https://scontent.cdninstagram.com/hphotos-xaf1...    4313\r\r likes   
1  https://scontent.cdninstagram.com/hphotos-xap1...    2102\r\r likes   
2  https://scontent.cdninstagram.com/hphotos-xap1...  108939\r\r likes   
3  https://scontent.cdninstagram.com/hphotos-xaf1...    6441\r\r likes   
4  https://scontent.cdninstagram.com/hphotos-xap1...    1153\r\r likes   

                                          owner_cmnt  \
0  gosh we love these five idiots so much __ for...   
1  Anyone see the Pepsi halftime show commercial ...   
2  Today I ate cr_pes and bought a drawing of ch...   
3  New works in progress for an upcoming show at ...   
4  Due to overwhelming response, we have went ahe...   

                                      owner_id   question1  \
0     <font color="#0066CC">maliksbby</font>       noneAgg   
1  <font color="#0066CC">witneycarson</font>       noneAgg   
2   <font color="#0066CC">namelessgem</font>       noneAgg   
3          <font color="#0066CC">kaws</font>    aggression   
4      <font color="#0066CC">fivefour</font>       noneAgg   

  question1:confidence question2 question2:confidence shared media     label  
0               0.8124   noneBll               1.0000         1477    normal  
1               1.0000   noneBll               1.0000          761    normal  
2               0.8056   noneBll               0.8056          698    normal  
3               0.5982  bullying               0.5982         1257  bullying  
4               1.0000   noneBll               1.0000         2052    normal  

[5 rows x 216 columns]

In [92]:
total_samples_in_dataset = len(final_df)
total_samples_in_dataset

2218

Finding image name for each row using mapping links with ids

In [93]:
image_id_0_10,image_url_0_10 = get_imageid_link_lists('data/instagram_images/namesall_info_media_ids_to_collect_0plus_to_10.txt')
image_id_10_40,image_url_10_40 = get_imageid_link_lists('data/instagram_images/namesall_info_media_ids_to_collect_10plus_to_40.txt')
image_id_40,image_url_40 = get_imageid_link_lists('data/instagram_images/namesall_info_media_ids_to_collect_40plus.txt')

In [94]:
image_ids = image_id_0_10+image_id_10_40+image_id_40
image_urls = image_url_0_10+image_url_10_40+image_url_40

In [95]:
len(image_ids),len(image_urls)

(1921, 1921)

In [96]:
image_urls = [i.split('/')[-1] for i in image_urls]

In [97]:
link_id_mapping_dict = dict(zip(image_urls,image_ids))

In [98]:
def return_id_from_link(row):
    try:
        return link_id_mapping_dict[row['img_url'].split('/')[-1]]
    except:
        return None

In [99]:
final_df['image_names_url_mapping'] = final_df.apply (lambda row: return_id_from_link(row),axis=1)

In [ ]:
final_df.to_pickle('data/DATA_I_MADE/final_csv_with_url_mapping_results.pkl')

In [10]:
final_df = pd.read_pickle('data/DATA_I_MADE/final_csv_with_url_mapping_results.pkl')

In [11]:
ids_retrived_using_links = [i for i in final_df['image_names_url_mapping'].tolist() if i!=None]
len(ids_retrived_using_links)

544

In [12]:
len(set(ids_retrived_using_links)&set(all_ids_folder))

543

That did not work out as expected so resolving to jugad

# #Jugad

Finding all he media_ids_present in data dump..

creating a id to dump file mapping of format 
```JSON
{
 id:dumpfile 
 .
 .
   
}
```

In [413]:
def fetch_all_media_ids_in_user_info_file(filename):
    with open(filename,'r') as reader:
        lines = reader.readlines()
        
    index = [i+1 for i, j in enumerate(lines) if j == 'media ID:\n']
    ids = [lines[i].rstrip() for i in index]
    return ids

In [420]:
normal_users_filenames = listdir('data/Instagram_normal_users/')
normal_users_filenames.remove('README.txt')
common_users_filenames=listdir('data/Instagram_common_users/')
common_users_filenames.remove('uesrid_common_insta_collected.txt')

In [421]:
len(normal_users_filenames),len(common_users_filenames)

(36, 11)

In [416]:
all_ids_data_dump = {}

In [417]:
for file in tqdm(normal_users_filenames):
    for id_ in fetch_all_media_ids_in_user_info_file(f'data/Instagram_normal_users/{file}'):
        if id_ in all_ids_data_dump:
            all_ids_data_dump[id_].append(file)
        else:
            all_ids_data_dump[id_] = [file]

In [422]:
for file in tqdm(common_users_filenames):
    for id_ in fetch_all_media_ids_in_user_info_file(f'data/Instagram_common_users/{file}'):
        if id_ in all_ids_data_dump:
            all_ids_data_dump[id_].append(file)
        else:
            all_ids_data_dump[id_] = [file]

In [423]:
with open(f'data/DATA_I_MADE/datadump_id_to_filename_map.json','w') as fp:
            json.dump(all_ids_data_dump, fp,indent=4)

In [424]:
with open(f'data/DATA_I_MADE/datadump_id_to_filename_map.json','r') as fp:
           all_ids_data_dump=json.load(fp)

In [425]:
len(all_ids_data_dump.keys())

5763354

In [426]:
len(all_ids_folder)

2263

In [427]:
ids_present_in_data_dump = list(set(all_ids_data_dump.keys())&set(all_ids_folder))
len(ids_present_in_data_dump)

1630

In [428]:
ids_not_present_in_data_dump = list(set(all_ids_folder)-set(ids_present_in_data_dump))
len(ids_not_present_in_data_dump)

633

In [18]:
ids_not_in_dump_but_retrived_using_links = set(ids_not_present_in_data_dump)&set(ids_retrived_using_links)
len(ids_not_in_dump_but_retrived_using_links)

227

## Parsing user datadump 

In [111]:
def read_user_datadump(filename):
    with open(filename,'r') as reader:
        lines = reader.readlines()
    return lines

In [112]:
def get_user_blocks(lines):
    user_blocks = []
    index_of_user_blocks=[i for i, j in enumerate(lines) if  "instagram user id:" in j] 
    number_of_blocks = len(index_of_user_blocks)
    for block_number, block_index in enumerate(index_of_user_blocks):
        if (block_number+1)==number_of_blocks:
            block_number-=1
        user_blocks.append(lines[block_index:index_of_user_blocks[block_number+1]])
    return user_blocks

In [153]:
def get_username_from_user_block(block):
    return block[0][block[0].index("user name:"):].split(':')[-1].strip()

In [162]:
def clean_comment_data_dump(comment):
    return ' '.join(comment.encode('ascii', 'ignore').strip( ' _'.encode('ascii')).decode("utf-8").split())

In [163]:
def get_media_id_and_comments_tuples(block):
    #Puling out all media blocks in user block
    index_media_blocks = [i for i, j in enumerate(block) if j == 'media ID:\n']
    number_of_media_blocks = len(index_media_blocks)
    media_blocks = []
    for block_number, block_index in enumerate(index_media_blocks):
        if (block_number+1)==number_of_media_blocks:
            block_number-=1
        media_blocks.append(block[block_index:index_media_blocks[block_number+1]])
    
    #making (comment,media_id) for each media block
    comment_id_tuples = []
    for m_block in media_blocks:
        if m_block!=[]:
            media_id_block = m_block[1].strip()
            comments_index = m_block.index('comments for this media:\n')
            like_index = m_block.index('likes for this media:\n')
            if like_index==comments_index+1:
                comment_id_tuples.append((None,media_id_block))     
            else:
                comment_lines_index = [i for i in range(comments_index+1,like_index) if "commentText:" in m_block[i]]
                for i in range(len(comment_lines_index)):
                    current_comment_index = comment_lines_index[i]
                    if i+1==len(comment_lines_index):
                        next_comment_index=like_index
                    else:
                        next_comment_index = comment_lines_index[i+1]
                    
                    comment_text = m_block[current_comment_index][m_block[current_comment_index].index('commentText:')+len('commentText:'):].strip()
                    
                    for i in range(current_comment_index+1,next_comment_index):
                        comment_text += m_block[i].strip() 
                    comment_text = clean_comment_data_dump(comment_text)    
                    comment_id_tuples.append((comment_text,media_id_block))                    
    return comment_id_tuples

In [164]:
all_files_to_parse = [f'data/Instagram_normal_users/{i}' for i in normal_users_filenames]+ [f'data/Instagram_common_users/{i}' for i in common_users_filenames]

Making Json from data dump..

Format
``` JSON
'username':{

    'comment1':media_id of tht comment
    .
    .
    .
}```

In [172]:
for file in tqdm(all_files_to_parse):
    username_comment_media_id_dict = dict()
    lines = read_user_datadump(file)
    user_blocks = [i for i in get_user_blocks(lines) if i!=[]]
    for user_block in user_blocks:
        username = get_username_from_user_block(user_block)
        media_id_comment_tuples = get_media_id_and_comments_tuples(user_block)
        temp_dict = dict(media_id_comment_tuples)
        if username in username_comment_media_id_dict.keys():
            username_comment_media_id_dict[username].update(temp_dict)
        else:
            username_comment_media_id_dict[username]=temp_dict
    dump_file_name = file.split('/')[-1].replace('.txt','')
    with open(f'data/DATA_I_MADE/username_indexed_datadump_jsons/{dump_file_name}.json','w') as fp:
        json.dump(username_comment_media_id_dict, fp,indent=4)
    count+=1

Making Json from data dump..

Format
``` JSON
{
'media_id':{
    'username': username of media owner
    'comments': [list of comments on that media]}
.
.
.
}```

In [173]:
for file in tqdm(all_files_to_parse):
    media_id_and_metadata_dict = dict()
    lines = read_user_datadump(file)
    user_blocks = [i for i in get_user_blocks(lines) if i!=[]]
    for user in user_blocks:
        username = get_username_from_user_block(user)
        media_id_tuples = get_media_id_and_comments_tuples(user)
        
        for comment_text, media_id in media_id_tuples:
            if media_id in media_id_and_metadata_dict.keys():
                media_id_and_metadata_dict[media_id]['comments'].append(comment_text)
            else:
                media_id_and_metadata_dict[media_id] = {'username':username,'comments':[comment_text]}
    dump_file_name = file.split('/')[-1].replace('.txt','')
    with open(f'data/DATA_I_MADE/media_id_indexed_datadump_jsons/{dump_file_name}.json','w') as fp:
        json.dump(media_id_and_metadata_dict, fp,indent=4)
    count+=1

creating username to datadump json filename mapping

In [174]:
filenames = listdir('data/DATA_I_MADE/username_indexed_datadump_jsons/')

In [175]:
username_file = dict()

In [176]:
for i in tqdm(filenames):
    with open(f'data/DATA_I_MADE/username_indexed_datadump_jsons/{i}','r') as reader:
        file_dict = json.load(reader)
    for user in file_dict.keys():
        if user in username_file.keys():
            username_file[user].append(i)
        else:
            username_file[user] = [i]

In [177]:
with open(f'data/DATA_I_MADE/username_to_file_mapping.json','w') as fp:
            json.dump(username_file, fp,indent=4)

creating media id to datadump json filename mapping

In [178]:
filenames = listdir('data/DATA_I_MADE/media_id_indexed_datadump_jsons/')

In [179]:
media_id_filename_mapping = dict()

In [180]:
for i in tqdm(filenames):
    with open(f'data/DATA_I_MADE/media_id_indexed_datadump_jsons/{i}','r') as reader:
        file_dict = json.load(reader)
    for media_id in file_dict.keys():
        if media_id in media_id_filename_mapping.keys():
            media_id_filename_mapping[media_id].append(i)
        else:
            media_id_filename_mapping[media_id] = [i]

In [181]:
with open(f'data/DATA_I_MADE/media_id_filename_mapping.json','w') as fp:
            json.dump(media_id_filename_mapping, fp,indent=4)

# Finding image for each sample

Helper functions

In [43]:
def clean_comment(comment):
    try:
        return comment[comment.index('</font>')+len('</font>'):comment.index('(created_at:')].replace('<font color="#0066CC">','').replace('</font>','').encode('ascii', 'ignore').strip( ' _'.encode('ascii')).decode("utf-8")
    except:
        return comment[comment.index('</font>')+len('</font>'):comment.index('(created at:')].replace('<font color="#0066CC">','').replace('</font>','').encode('ascii', 'ignore').strip( ' _'.encode('ascii')).decode("utf-8")

## method 1

For each row:
    1. Extract comments present in row -> Clean comment
    2. Extract username of owner
    3. Extract files in which that user block is present
    4. Extract extract user block
    5. Find out ids by matching comments extracted in point 1

In [183]:
with open('data/DATA_I_MADE/username_to_file_mapping.json','r') as reader:
    username_file_map = json.load(reader)

In [184]:
final_df

_golden _last_judgment_at  _trusted_judgments   _unit_id _unit_state  \
0       False      4/11/15 0:40                   5  702714440   finalized   
1       False      4/11/15 0:37                   5  702714441   finalized   
2       False      4/11/15 2:17                   5  702714442   finalized   
3       False      4/11/15 2:09                   5  702714443   finalized   
4       False     4/10/15 23:41                   5  702714444   finalized   
5       False     4/10/15 23:44                   5  702714445   finalized   
6       False      4/11/15 0:40                   5  702714446   finalized   
7       False      4/11/15 0:37                   5  702714447   finalized   
8       False     4/10/15 23:48                   5  702714448   finalized   
9       False     4/10/15 23:47                   5  702714449   finalized   
10      False      4/11/15 0:42                   5  702714450   finalized   
11      False      4/11/15 2:10                   5  702714451   finalized   
12      False     4/10/15 23:48                   5  702714452   finalized   
13      False      4/11/15 2:09                   5  702714453   finalized   
14      False      4/11/15 2:11                   5  702714454   finalized   
15      False     4/10/15 23:45                   5  702714455   finalized   
16      False     4/10/15 23:44                   5  702714456   finalized   
17      False     4/10/15 23:41                   5  702714457   finalized   
18      False      4/11/15 2:14                   5  702714458   finalized   
19      False      4/11/15 2:10                   5  702714459   finalized   
20      False      4/11/15 2:08                   5  702714460   finalized   
21      False      4/11/15 2:07                   5  702714461   finalized   
22      False     4/10/15 23:48                   5  702714462   finalized   
23      False     4/10/15 23:44                   5  702714463   finalized   
24      False      4/11/15 2:14                   5  702714464   finalized   
25      False      4/11/15 2:17                   5  702714465   finalized   
26      False     4/10/15 23:34                   5  702714466   finalized   
27      False     4/10/15 23:35                   5  702714467   finalized   
28      False     4/10/15 23:37                   5  702714468   finalized   
29      False      4/11/15 0:42                   5  702714469   finalized   
...       ...               ...                 ...        ...         ...   
2188    False     1/22/15 20:50                   5  652910847   finalized   
2189    False     1/20/15 16:33                   5  652910848   finalized   
2190    False      1/24/15 4:20                   5  652910849   finalized   
2191    False      1/22/15 0:49                   5  652910850   finalized   
2192    False      1/21/15 6:16                   5  652910851   finalized   
2193    False     1/23/15 11:39                   5  652910852   finalized   
2194    False     1/23/15 11:47                   5  652910853   finalized   
2195    False     1/27/15 14:30                   5  652910854   finalized   
2196    False     1/22/15 21:19                   5  652910855   finalized   
2197    False      1/26/15 0:03                   5  652910856   finalized   
2198    False     1/20/15 17:43                   5  652910857   finalized   
2199    False     1/24/15 18:21                   5  652910858   finalized   
2200    False     1/23/15 11:50                   5  652910859   finalized   
2201    False      1/22/15 1:59                   5  652910860   finalized   
2202    False      1/26/15 0:08                   5  652910861   finalized   
2203    False     1/27/15 14:21                   5  652910862   finalized   
2204    False     1/20/15 12:48                   5  652910863   finalized   
2205    False     1/21/15 17:24                   5  652910864   finalized   
2206    False     1/23/15 11:44                   5  652910865   finalized   
2207    False      1/27/15 2

In [190]:
count = 0

In [193]:
def get_image_id(row):
    comments = [row[f'clmn{i}'] for i in range(1,196)]
    comments = [clean_comment(i) for i in comments if i != 'empety']
    username  = row['owner_id'].replace("</font>","").replace('<font color="#0066CC">','').strip()
    
    if username in username_file_map.keys():
        mapping_file_number = username_file_map[username]
    else:    
        return None

    
    ids=[]
    for filename in mapping_file_number:
        with open(f'data/DATA_I_MADE/username_indexed_datadump_jsons/{filename}','r') as reader:
            file_json = json.load(reader)
        user_block = file_json[username]
                
        for comment in comments:
            try:
                ids.append(user_block[comment])
            except:
                continue
    
    if ids!=[]:
        return max(set(ids), key = ids.count)
    else:
        return None

In [199]:
# final_df['image_names_juggad_method_1'] = final_df.apply (lambda row: get_image_id(row),axis=1)
final_df['image_names_juggad_method_1'] = final_df.progress_apply (lambda row: get_image_id(row),axis=1)

In [200]:
final_df

_golden _last_judgment_at  _trusted_judgments   _unit_id _unit_state  \
0       False      4/11/15 0:40                   5  702714440   finalized   
1       False      4/11/15 0:37                   5  702714441   finalized   
2       False      4/11/15 2:17                   5  702714442   finalized   
3       False      4/11/15 2:09                   5  702714443   finalized   
4       False     4/10/15 23:41                   5  702714444   finalized   
5       False     4/10/15 23:44                   5  702714445   finalized   
6       False      4/11/15 0:40                   5  702714446   finalized   
7       False      4/11/15 0:37                   5  702714447   finalized   
8       False     4/10/15 23:48                   5  702714448   finalized   
9       False     4/10/15 23:47                   5  702714449   finalized   
10      False      4/11/15 0:42                   5  702714450   finalized   
11      False      4/11/15 2:10                   5  702714451   finalized   
12      False     4/10/15 23:48                   5  702714452   finalized   
13      False      4/11/15 2:09                   5  702714453   finalized   
14      False      4/11/15 2:11                   5  702714454   finalized   
15      False     4/10/15 23:45                   5  702714455   finalized   
16      False     4/10/15 23:44                   5  702714456   finalized   
17      False     4/10/15 23:41                   5  702714457   finalized   
18      False      4/11/15 2:14                   5  702714458   finalized   
19      False      4/11/15 2:10                   5  702714459   finalized   
20      False      4/11/15 2:08                   5  702714460   finalized   
21      False      4/11/15 2:07                   5  702714461   finalized   
22      False     4/10/15 23:48                   5  702714462   finalized   
23      False     4/10/15 23:44                   5  702714463   finalized   
24      False      4/11/15 2:14                   5  702714464   finalized   
25      False      4/11/15 2:17                   5  702714465   finalized   
26      False     4/10/15 23:34                   5  702714466   finalized   
27      False     4/10/15 23:35                   5  702714467   finalized   
28      False     4/10/15 23:37                   5  702714468   finalized   
29      False      4/11/15 0:42                   5  702714469   finalized   
...       ...               ...                 ...        ...         ...   
2188    False     1/22/15 20:50                   5  652910847   finalized   
2189    False     1/20/15 16:33                   5  652910848   finalized   
2190    False      1/24/15 4:20                   5  652910849   finalized   
2191    False      1/22/15 0:49                   5  652910850   finalized   
2192    False      1/21/15 6:16                   5  652910851   finalized   
2193    False     1/23/15 11:39                   5  652910852   finalized   
2194    False     1/23/15 11:47                   5  652910853   finalized   
2195    False     1/27/15 14:30                   5  652910854   finalized   
2196    False     1/22/15 21:19                   5  652910855   finalized   
2197    False      1/26/15 0:03                   5  652910856   finalized   
2198    False     1/20/15 17:43                   5  652910857   finalized   
2199    False     1/24/15 18:21                   5  652910858   finalized   
2200    False     1/23/15 11:50                   5  652910859   finalized   
2201    False      1/22/15 1:59                   5  652910860   finalized   
2202    False      1/26/15 0:08                   5  652910861   finalized   
2203    False     1/27/15 14:21                   5  652910862   finalized   
2204    False     1/20/15 12:48                   5  652910863   finalized   
2205    False     1/21/15 17:24                   5  652910864   finalized   
2206    False     1/23/15 11:44                   5  652910865   finalized   
2207    False      1/27/15 2

In [201]:
final_df.to_pickle('data/DATA_I_MADE/final_csv_with_juggad_method_1.pkl')

Looking into found ids

In [378]:
final_df = pd.read_pickle('data/DATA_I_MADE/final_csv_with_juggad_method_1.pkl')

In [379]:
image_names_list = final_df['image_names_juggad_method_1'].tolist()

In [380]:
ids_found = [i for i in image_names_list if i!=None]
len(ids_found)

1444

In [381]:
print(f'unique ids found {len(set(ids_found))}')

unique ids found 1432


In [382]:
duplicate_ids_method_1 = []

In [383]:
result = sorted([(x, ids_found.count(x)) for x in set(ids_found)], key=lambda y: y[1])
for elem in result:
    print(f'{elem[0]} {elem[1]}')
    if elem[1]>1:
        duplicate_ids_method_1.append(elem[0])

146661900656090644_8328110 1
199232569673006095_5045782 1
340819075707144954_193173297 1
514183310239465360_322644222 1
408195621997913047_285955903 1
728603711082761829_6341954 1
461800630787484056_15982281 1
314573507749860817_13307932 1
679125106333801970_336331238 1
174871033487041154_20315007 1
399433708552404926_304296358 1
555642493363190956_298930414 1
499633003736202929_322644222 1
454320667353994039_206604147 1
659385455921847806_1269598 1
724895937495131607_224350344 1
287591731421938871_31254000 1
259006092381101142_45755614 1
658633856583479375_12853339 1
223416251026773484_187325004 1
421248719213279441_7519845 1
496547915259905864_4145448 1
587700605248789969_21832230 1
707211097157054073_22869266 1
479103988997976268_145598227 1
600945564806541876_293847299 1
426084081563561911_7565730 1
556326291008210398_28939176 1
531858247107595649_1783442 1
579212890660313714_4602213 1
392839494392456028_145125103 1
641647492756922614_218144855 1
179253843383574349_28505022 1
36322

In [384]:
duplicate_ids_method_1

['198765473015075272_54217883',
 '651674909315005888_369159190',
 '416797763844580918_292277184',
 '168700116936286807_29930887',
 '559239604827122306_397255825',
 '647173406669644021_379452023',
 '500767512191447403_460563723',
 '643045993521867241_278236785',
 '227422412880765978_182393608']

In [305]:
intersection_ids_found_and_ids_in_folder = set(all_ids_folder)&set(ids_found)
len(intersection_ids_found_and_ids_in_folder)

742

In [306]:
len(set(ids_present_in_data_dump)&set(ids_found))

742

In [307]:
len(set(ids_retrived_using_links)&set(ids_found))

285

In [27]:
# label_list = final_df.loc[final_df['image_names_juggad_method_1'].isin(intersection_ids_found_and_ids_in_folder)]['label'].tolist()

## Method 2

For each image in folder:
    1. If present in data dump extract its username and comments
    2. Extract comments for all samples of that username in labeled data
    3. Match comments to find the correct labeled sample

Converting labbeled CSV into JSON for faster search

In [140]:
labbeled_data_formatted = dict()

In [141]:
def extract_to_dict(row):
    comments = [row[f'clmn{i}'] for i in range(1,196)]
    comments = [clean_comment(i) for i in comments if i != 'empety']
    unit_id = row['_unit_id']
    
    temp = {}
    for comment in comments:
        temp[comment] = unit_id
        
    return temp

In [142]:
all_dicts = final_df.progress_apply(lambda row: extract_to_dict_try(row),axis=1);

In [143]:
for i in all_dicts:
    for j in i.keys():
        if j in labbeled_data_formatted.keys():
            labbeled_data_formatted[j].append(i[j])
        else:
            labbeled_data_formatted[j] = [i[j]]

In [145]:
with open(f'data/DATA_I_MADE/label_data_extracted.json','w') as fp:
            json.dump(labbeled_data_formatted, fp,indent=4)

matching now

In [252]:
with open(f'data/DATA_I_MADE/media_id_filename_mapping.json','r') as fp:
           media_id_file=json.load(fp)
        
with open(f'data/DATA_I_MADE/label_data_extracted.json','r') as fp:
           labbeled_data_formatted=json.load(fp)

In [253]:
len(all_ids_folder)

2263

In [287]:
media_id_unit_id_map = {}

In [255]:
file_names = listdir('data/DATA_I_MADE/media_id_indexed_datadump_jsons/')
all_dicts_loaded = {}
for i in tqdm(file_names):
    with open(f'data/DATA_I_MADE/media_id_indexed_datadump_jsons/{i}','r') as fp:
            all_dicts_loaded[i] = json.load(fp)

In [288]:
for media_id in tqdm(all_ids_folder):
    try:
        files_to_read = media_id_file[media_id]
    except:
        continue
    
    unit_ids = []
    for file in files_to_read:
#         with open(f'data/DATA_I_MADE/media_id_indexed_datadump_jsons/{file}','r') as fp:
#             all_id_blocks = json.load(fp)
        all_id_blocks = all_dicts_loaded[file]
        
        id_block = all_id_blocks[media_id]
        comments_data_dump = id_block['comments']
        
        comments_data_dump = [i for i in comments_data_dump if i != '']
        comments_data_dump = [i for i in comments_data_dump if len(i.split())>1]
        
        for comment in comments_data_dump:
            try:
                unit_ids.extend(labbeled_data_formatted[comment])
            except:
                continue
    
    if unit_ids!=[]:
        if len(set(unit_ids)) == 1:
            media_id_unit_id_map[media_id] = unit_ids[0]
        else:
            result = sorted([(x, unit_ids.count(x)) for x in set(unit_ids)], key=lambda y: y[1])
            if result[-1][1] > result[-2][1]:
                media_id_unit_id_map[media_id] = result[-1][0]

In [289]:
val = list(media_id_unit_id_map.values())
len(val),len(set(val))

(943, 872)

Lets fix this duplicate issue first....

After checking I found (Checking removed to keep the code clean)

This occurs as some images have suset of comments common with other image and no/ not enough unique comment of there own...

Simple solution...map with id which have most matching comments and delete rest entries for now...

In [290]:
result = sorted([(x, val.count(x)) for x in set(val)], key=lambda y: y[1])

In [291]:
ids_in_question  = [i[0] for i in result if i[1]>1]

In [292]:
for id_in_question in tqdm(ids_in_question):
    media_ids_questioned = [k for k in media_id_unit_id_map.keys() if media_id_unit_id_map[k] == id_in_question]
    
    media_hit_count_map = {}
    for media_id in media_ids_questioned:
        try:
            files_to_read = media_id_file[media_id]
        except:
            continue

        unit_ids = []
        for file in files_to_read:
            all_id_blocks = all_dicts_loaded[file]

            id_block = all_id_blocks[media_id]
            comments_data_dump = id_block['comments']

            comments_data_dump = [i for i in comments_data_dump if i != '']
            comments_data_dump = [i for i in comments_data_dump if len(i.split())>1]

            for comment in comments_data_dump:
                try:
                    unit_ids.extend(labbeled_data_formatted[comment])
                except:
                    continue
        if unit_ids!=[]:
            media_hit_count_map[media_id] = unit_ids.count(id_in_question)
    max_key = max(media_hit_count_map.items(), key=operator.itemgetter(1))[0]
    media_hit_count_map.pop(max_key,None)
    for k in media_hit_count_map.keys():
        media_id_unit_id_map.pop(k,None)

In [311]:
ids_found = list(media_id_unit_id_map.keys())

In [294]:
len(ids_found),len(ids_present_in_data_dump)

(872, 1630)

In [295]:
still_not_found = set(ids_present_in_data_dump)-set(ids_found)
len(still_not_found)

758

In [296]:
val = list(media_id_unit_id_map.values())
len(val),len(set(val))

(872, 872)

In [308]:
len(set(ids_retrived_using_links)&set(ids_found))

285

In [313]:
len(intersection_ids_found_and_ids_in_folder-set(ids_found))

4

In [297]:
with open(f'data/DATA_I_MADE/media_id_unit_id_mapping_method2_results.json','w') as fp:
            json.dump(media_id_unit_id_map, fp,indent=4)

Duplicates issue resolved..Now time to find why I am not able to map some ids which are present in the dump 

After manuall incpection it looks like that rows for those images are not present in dataset....It should not be the case...will look into it later...

Dumping found ids from method 2 into Final df

In [388]:
final_df['image_names_juggad_method_2'] = None

In [389]:
for key, value in media_id_unit_id_map.items():
    index = final_df.index[final_df['_unit_id'] == value].tolist()[0]
    final_df.loc[index,'image_names_juggad_method_2'] = key

In [399]:
final_df

_golden _last_judgment_at  _trusted_judgments   _unit_id _unit_state  \
0       False      4/11/15 0:40                   5  702714440   finalized   
1       False      4/11/15 0:37                   5  702714441   finalized   
2       False      4/11/15 2:17                   5  702714442   finalized   
3       False      4/11/15 2:09                   5  702714443   finalized   
4       False     4/10/15 23:41                   5  702714444   finalized   
5       False     4/10/15 23:44                   5  702714445   finalized   
6       False      4/11/15 0:40                   5  702714446   finalized   
7       False      4/11/15 0:37                   5  702714447   finalized   
8       False     4/10/15 23:48                   5  702714448   finalized   
9       False     4/10/15 23:47                   5  702714449   finalized   
10      False      4/11/15 0:42                   5  702714450   finalized   
11      False      4/11/15 2:10                   5  702714451   finalized   
12      False     4/10/15 23:48                   5  702714452   finalized   
13      False      4/11/15 2:09                   5  702714453   finalized   
14      False      4/11/15 2:11                   5  702714454   finalized   
15      False     4/10/15 23:45                   5  702714455   finalized   
16      False     4/10/15 23:44                   5  702714456   finalized   
17      False     4/10/15 23:41                   5  702714457   finalized   
18      False      4/11/15 2:14                   5  702714458   finalized   
19      False      4/11/15 2:10                   5  702714459   finalized   
20      False      4/11/15 2:08                   5  702714460   finalized   
21      False      4/11/15 2:07                   5  702714461   finalized   
22      False     4/10/15 23:48                   5  702714462   finalized   
23      False     4/10/15 23:44                   5  702714463   finalized   
24      False      4/11/15 2:14                   5  702714464   finalized   
25      False      4/11/15 2:17                   5  702714465   finalized   
26      False     4/10/15 23:34                   5  702714466   finalized   
27      False     4/10/15 23:35                   5  702714467   finalized   
28      False     4/10/15 23:37                   5  702714468   finalized   
29      False      4/11/15 0:42                   5  702714469   finalized   
...       ...               ...                 ...        ...         ...   
2188    False     1/22/15 20:50                   5  652910847   finalized   
2189    False     1/20/15 16:33                   5  652910848   finalized   
2190    False      1/24/15 4:20                   5  652910849   finalized   
2191    False      1/22/15 0:49                   5  652910850   finalized   
2192    False      1/21/15 6:16                   5  652910851   finalized   
2193    False     1/23/15 11:39                   5  652910852   finalized   
2194    False     1/23/15 11:47                   5  652910853   finalized   
2195    False     1/27/15 14:30                   5  652910854   finalized   
2196    False     1/22/15 21:19                   5  652910855   finalized   
2197    False      1/26/15 0:03                   5  652910856   finalized   
2198    False     1/20/15 17:43                   5  652910857   finalized   
2199    False     1/24/15 18:21                   5  652910858   finalized   
2200    False     1/23/15 11:50                   5  652910859   finalized   
2201    False      1/22/15 1:59                   5  652910860   finalized   
2202    False      1/26/15 0:08                   5  652910861   finalized   
2203    False     1/27/15 14:21                   5  652910862   finalized   
2204    False     1/20/15 12:48                   5  652910863   finalized   
2205    False     1/21/15 17:24                   5  652910864   finalized   
2206    False     1/23/15 11:44                   5  652910865   finalized   
2207    False      1/27/15 2

# Getting final image name for row

---
**NOTE**

Discarding results of method 1 here because...

only 4 images which are not found by link/method 2 and are present in folder are found.. out of 4, 2 are duplicate and other 2 conflict with results of method 2/link..

As method 2 and Link are more realiable, disregarding method 1

---

In [394]:
def get_final_image_name(row):
    if row['image_names_url_mapping'] !=None:
        return row['image_names_url_mapping']
    
    if row['image_names_juggad_method_2'] !=None:
        return row['image_names_juggad_method_2']
    
    return None

In [395]:
final_df['final_image_name'] = final_df.progress_apply (lambda row: get_final_image_name(row),axis=1)

In [400]:
final_df.to_pickle('data/DATA_I_MADE/final_csv_with_image_names.pkl')

In [396]:
images_found = [i for i in final_df['final_image_name'] if i !=None]
len(images_found),len(set(images_found))

(1099, 1099)

In [397]:
labels = [j for i,j in zip(final_df['final_image_name'],final_df['label']) if i !=None]
len(labels),len(set(labels))

(1099, 3)

In [401]:
result = sorted([(x, labels.count(x)) for x in set(labels)], key=lambda y: y[1])
for elem in result:
    print(f'{elem[0]} {elem[1]}')

aggression 117
bullying 440
normal 542


In [472]:
number_of_none = final_df['final_image_name'].tolist().count(None)
number_of_none

1119

In [474]:
image_not_found = [i for i in all_ids_folder if i not in images_found]

In [478]:
len(image_not_found),len(images_found),len(image_not_found)+len(images_found),len(all_ids_folder)

(1165, 1099, 2264, 2263)

In [482]:
present_in_dump_but_not_found=set(ids_present_in_data_dump)&set(image_not_found)
len(present_in_dump_but_not_found)

759

# Checking why I can not find rest images match

In [522]:
count=0

In [525]:
for media_id in tqdm(present_in_dump_but_not_found):
    try:
        files_to_read = media_id_file[media_id]
    except:
        continue
    
    unit_ids = []
    matching_comments = []
    for file in files_to_read:
#         with open(f'data/DATA_I_MADE/media_id_indexed_datadump_jsons/{file}','r') as fp:
#             all_id_blocks = json.load(fp)
        all_id_blocks = all_dicts_loaded[file]
        
        id_block = all_id_blocks[media_id]
        comments_data_dump = id_block['comments']
        
        comments_data_dump = [i for i in comments_data_dump if i != '']
        comments_data_dump = [i for i in comments_data_dump if len(i.split())>1]
        
        
        for comment in comments_data_dump:
            try:
                unit_ids.extend(labbeled_data_formatted[comment])
                matching_comments.append(comment)
            except:
                continue
    print(f'-------------{media_id}----------------')
    if unit_ids == []:
        count+=1
        print(len(comments_data_dump))
        print(comments_data_dump)
    else:
        
        if len(set(unit_ids)) == 1:
                print(unit_ids,len(unit_ids))
                print(matching_comments)
        else:
            result = sorted([(x, unit_ids.count(x)) for x in set(unit_ids)], key=lambda y: y[1])
            print(f'{result[-1][1]}  {result[-2][1]}')
            print(matching_comments)

-------------500829360775245650_208206546----------------
1  1
['Luv it!', 'Luv it']
-------------667634612782830206_54217883----------------
93
['Yesss @clara_wells', "I hope you don't mind me spamming you", "You're never gonna notice me anyways so lmfao @jackgilinsky", '@baby_meggie we need a pic like this', '@altheaedang YESSS !', "@baby_meggie HAHAHA I'm serious about it so much we like really NEED one", '@altheaedang if we can make the faces', "@baby_meggie LMAO. If not, I'll find another one", '0_0 < @shimalaa', "@olivia_mcc hayes's shirt", '@espinosians gave me his snap', '@nat_walsh88 Cameron Dallas is my boyfriendddddd -shawnie @nat_walsh88', "@olivia_mcc I'm sitting next to a pile of lemons come on over its really a blast", "HAHAHAHAHAHAHA that's so normal do u have escuela tomorrow @nat_walsh88", '@olivia_mcc yes because this country wants me to become #stoopid', '@nat_walsh88 wuts 9+10 ?????', '@olivia_mcc twenny one', 'U STOOPID @nat_walsh88', '@magcon_numbersreal @magcon_

1  1
['Amazing !', 'Love this!!']
-------------719008428690199023_421656491----------------
1  1
['Fucking perfection', 'Very sexy', 'Troco likes']
-------------651058834177629950_45755614----------------
[702714443] 1
['Need one']
-------------694409359246207661_41526257----------------
1  1
['First like']
-------------440004345775942457_20861689----------------
15
['You my #mancrushmonday', "What's good my nigga", 'Your just to good looking I love your music keep it up :) :) :) @jah_youthlopez', "That's because your hot haha", 'oh jinsu', '@jah_youthlopez @ricksteelmusic', "@ricksteelmusic what's good my niggs.. Miss you bro", 'Chilling bro when u back in nyc', 'Yo jin whats good...i need to holler at you biut some shit homie.', 'How you go reply to him n not your wife @jah_youthlopez', '@ricksteelmusic after this Floyd fight', 'THATS right', 'My baby is growing uppp. *tear* lol', 'I love u please let my get them digits real quick so finnnneeew', "Lol. I didn't know you had an IG Jah

In [524]:
f'No Comment match for {count} images'

'No Comment match for 374 images'

-----
Seems like 374 images have zero mattchin comments in labeled csv and rest have very few comments (for most part very common comments) matching with one or multiple rows in labbeled data but all these row has matching images.

However this raises a question...That does images present in folder and labeled in dataset are different?? 

-----

Finding users of images not found

In [544]:
image_not_found_users_count_map = dict()

In [545]:
for media_id in tqdm(present_in_dump_but_not_found):
    try:
        files_to_read = media_id_file[media_id]
    except:
        continue
    
    for file in files_to_read:
#         with open(f'data/DATA_I_MADE/media_id_indexed_datadump_jsons/{file}','r') as fp:
#             all_id_blocks = json.load(fp)
        all_id_blocks = all_dicts_loaded[file]
        id_block = all_id_blocks[media_id]
        username = id_block['username']
        
    if username in image_not_found_users_count_map.keys():
        image_not_found_users_count_map[username]['count']+=1
        image_not_found_users_count_map[username]['media_id'].append(media_id)
    else:
        image_not_found_users_count_map[username]={'count':1,'media_id':[media_id]}

In [546]:
image_not_found_users_count_map

{'hydracup': {'count': 1, 'media_id': ['500829360775245650_208206546']},
 'jackgilinsky': {'count': 2,
  'media_id': ['667634612782830206_54217883', '630008949255245831_54217883']},
 'dotzsoh': {'count': 1, 'media_id': ['248004416718034154_1843380']},
 'physicalnovel': {'count': 1, 'media_id': ['384644859569120345_23959872']},
 'nipplejousting': {'count': 2,
  'media_id': ['630050190849497891_205489117',
   '629233703003904473_205489117']},
 'justjroc': {'count': 1, 'media_id': ['568309385488793428_26971310']},
 'jaythewanted': {'count': 1, 'media_id': ['180470464242476537_43677008']},
 'epic_dimes': {'count': 1, 'media_id': ['695287935460626583_317679019']},
 'bobbytribal': {'count': 3,
  'media_id': ['255201749_3778225',
   '740548865143499431_3778225',
   '557451797253482867_3778225']},
 'joeygraceffa': {'count': 3,
  'media_id': ['700203066665062352_1783442',
   '326580743_1783442',
   '398613108940459520_1783442']},
 'stoked_yogi': {'count': 1, 'media_id': ['709822513978593926_233

Users in dataframe with image as None

In [551]:
labels_not_found_users_count_map=dict()

In [552]:
for user,media,u_id in zip(final_df['owner_id'],final_df['final_image_name'],final_df['_unit_id']):
    user = user.replace('<font color="#0066CC">','').replace('</font>','').strip()
    if media == None:
        if user in labels_not_found_users_count_map.keys():
            labels_not_found_users_count_map[user]['count']+=1
            labels_not_found_users_count_map[user]['unit_id'].append(u_id)
        else:
            labels_not_found_users_count_map[user]={'count':1,'unit_id':[u_id]}

In [568]:
len(set(labels_not_found_users_count_map.keys())),len(set(image_not_found_users_count_map.keys()))

(540, 532)

In [572]:
len(set(labels_not_found_users_count_map.keys())&set(image_not_found_users_count_map.keys()))

170

In [573]:
for k in set(labels_not_found_users_count_map.keys())&set(image_not_found_users_count_map.keys()):
    if image_not_found_users_count_map[k]['count'] == labels_not_found_users_count_map[k]['count']==1:
        print(f"{k}----{image_not_found_users_count_map[k]['count']}-{image_not_found_users_count_map[k]['media_id']}------{labels_not_found_users_count_map[k]['count']}-{labels_not_found_users_count_map[k]['unit_id']}")

volcomskate----1-['576304132655507407_224350344']------1-[679605256]
steviewilliams----1-['627748657042900996_6474433']------1-[679605204]
hulyavsr----1-['604207282928461326_295097536']------1-[679604951]
theratchetmess----1-['549513752504796269_215303936']------1-[679604729]
factsdaily----1-['715053335689413745_352788696']------1-[679604925]
angelbespoke----1-['601329604442323024_3743453']------1-[679605271]
twerkingwithclouds----1-['683673821523279582_484182110']------1-[679605314]
channingtatum----1-['282856907654308084_20315007']------1-[679604424]
bonesbearings----1-['199186186928198702_29875324']------1-[679604651]
the1975----1-['531879221522204827_43153969']------1-[679605062]
djkidfamous----1-['589541807121322709_7858411']------1-[679605246]
surfingmagazine----1-['494074606260512805_10173257']------1-[679605129]
trendablapp----1-['526071773697248257_281461378']------1-[679604474]
theevanitydiary----1-['688132516074559003_220324']------1-[679605002]
mrmauricio----1-['16800619297

On manuall checking I found that matching users may not work always...I Have to manually check each one..too much work for now

# Moving Images

In [586]:
def get_folder_name(image_name):
    if image_name in image_id_0_10_folder:
        return 'data/instagram_images/images_0_10/'
    elif image_name in image_id_10_40_folder:
        return 'data/instagram_images/images_10_40/'
    else:
        return 'data/instagram_images/images_40plus/'

In [587]:
error_files = []
for img,l in tqdm(zip(final_df['final_image_name'],final_df['label'])):
    if img != None:
        file_path = get_folder_name(img)
        try:
            if l== 'normal':
                copyfile(f'{file_path}{img}.jpg', f'data/DATA_I_MADE/Images/Normal/{img}.jpg')
            else:
                copyfile(f'{file_path}{img}.jpg', f'data/DATA_I_MADE/Images/Trolled/{img}.jpg')
        except:
            error_files.append(img)

In [588]:
error_files

['702957319868735890_1262125599']

In [596]:
!echo "Number of Normal images"
!ls -l data/DATA_I_MADE/Images/Normal | egrep -c '^-'

Number of Normal images
541


In [597]:
!echo "Number of Trolled images"
!ls -l data/DATA_I_MADE/Images/Trolled | egrep -c '^-'

Number of Trolled images
557
